server/WFCORE-4251_list_modules_in_deployment_classpath.adoc

```diff
@@ -0,0 +1,118 @@
+= Provide ability to list modules which are on deployment's classpath via a Management Operation.
+:author:            Yeray Borges
+:email:             yborgess@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse/
+
+== Overview
+Currently, the list of modules available at classpath by each deployment is only available via jconsole:
+
+jboss.modules -> ModuleLoader -> ServiceModuleLoader -> Operations -> getDependencies( "deployment.<application_war_ear_name>" )
+
+This document describes the RFE to provide the ability to expose this information via a management operation.
+
+When optimizing the application footprint, this information becomes important:
+
+* The user could use this list to identify which dependencies are added implicitly by the server, and then, exclude any unneeded dependency from the deployment.
+* The user could use this information to discover any duplicated artifact bundled together with the deployment and included in one of the modules that the server is adding to the application classpath.
+* This information could facilitate discovering unneeded subsystems in the installation of his servers.
+
+
+The proposed solution is to include a new operation `list-modules` available under `/deployment` and, in case of ears, `/deployment=*/subdeployment` resource:
```

Comment:
Re: ears for each of the deployment or subdeployment resources, what is shown re: the other? IOW there are module dependencies between the elements of these kinds of deployments -- are these shown?

I don't think there should be any special magic handling for these, i.e. working out data that isn't available in a similar way to how you get other module dependency data. My question is more one of clarifying what can / cannot be displayed given the data that's available.

Reply:
> Re: ears for each of the deployment or sub-deployment resources, what is shown re: the other? IOW there are module dependencies between the elements of these kinds of deployments -- are these shown?

Yes, we will see the dependencies added implicitly by the server and the shared modules between them. E.g, if we deploy an ear that contains a war and an ejb-jar, when we are listing the modules of the war, we will see the modules added implicitly by the server plus the ear module and the ejb-jar module, these latter ones being modules created by the server at deployment time.

> I don't think there should be any special magic handling for these, i.e. working out data that isn't available in a similar way to how you get other module dependency data. My question is more one of clarifying what can / cannot be displayed given the data that's available.

We have access to the list of final modules created by each deployment and sub-deployment. We could even have the list split by modules added by the user, modules added by the server and the modules created by the server for the deployment file deployed by the user.

Reply:
Can you spell out these details in the proposal? What you say sounds good.

Reply:
I've just updated the PR with the output divided into dependency categories; system dependencies, user dependencies, and local dependencies